lib/image-editor/class-image-editor.php

```diff
@@ -0,0 +1,260 @@
+ */
+require_once __DIR__ . '/class-image-editor-flip.php';
+
+/**
+ * Image rotate modifier.
+ */
+require_once __DIR__ . '/class-image-editor-rotate.php';
+
+/**
+ * Image crop modifier.
+ */
+require_once __DIR__ . '/class-image-editor-crop.php';
+
+/**
+ * Image editor.
+ *
+ * @since 7.x ?
+ */
+class Image_Editor {
+	const META_KEY = 'image-editor';
+
+	/**
+	 * Constructs an Image_Editor.
+	 */
+	public function __construct() {
+		$this->all_modifiers = array(
+			'Image_Editor_Crop',
+			'Image_Editor_Flip',
+			'Image_Editor_Rotate',
+		);
+	}
+
+	/**
+	 * Modifies an image.
+	 *
+	 * @param integer $media_id Media id.
+	 * @param array   $modifiers Modifier to apply to the image.
+	 * @return array|WP_Error If successful image JSON containing the media_id and url of modified image, otherwise WP_Error.
+	 */
+	public function modify_image( $media_id, $modifiers ) {
+		// Get image information.
+		$info = $this->load_image_info( $media_id );
+		if ( is_wp_error( $info ) ) {
+			return $info;
+		}
+
+		foreach ( $modifiers as $modifier ) {
+			// Update it with our modifier.
+			$info['meta'] = $modifier->apply_to_meta( $info['meta'] );
+		}
+
+		// Generate filename based on current attributes.
+		$target_file = $this->get_filename( $info['meta'] );
+
+		// Does the image already exist?
+		$existing_image = $this->get_existing_image( $info, $target_file );
+		if ( $existing_image ) {
+			// Return the existing image.
+			return $existing_image;
+		}
+
+		// Try and load the image itself.
+		$image = $this->load_image( $media_id, $info );
+		if ( is_wp_error( $image ) ) {
+			return $image;
+		}
+
+		foreach ( $modifiers as $modifier ) {
+			// Finally apply the modification.
+			$modified = $modifier->apply_to_image( $image['editor'] );
+			if ( is_wp_error( $modified ) ) {
+				return $modified;
+			}
+		}
+
+		// And save.
+		return $this->save_image( $image, $target_file, $info );
+	}
+
+	/**
+	 * Loads an image for editing.
+	 *
+	 * @param integer $media_id Image ID.
+	 * @return array|WP_Error The WP_Image_Editor and image path if successful, WP_Error otherwise.
+	 */
+	private function load_image( $media_id ) {
+		require_once ABSPATH . 'wp-admin/includes/image.php';
+
+		$image_path = get_attached_file( $media_id );
+
+		if ( empty( $image_path ) ) {
+			return new WP_Error( 'rest_cannot_find_attached_file', __( 'Unable to find original media file.', 'gutenberg' ), array( 'status' => 500 ) );
+		}
+
+		$image_editor = wp_get_image_editor( $image_path );
+		if ( ! $image_editor->load() ) {
+			return new WP_Error( 'rest_cannot_load_editor', __( 'Unable to load original media file.', 'gutenberg' ), array( 'status' => 500 ) );
+		}
+
+		return array(
+			'editor' => $image_editor,
+			'path'   => $image_path,
+		);
+	}
+
+	/**
+	 * Gets the JSON response object for an image.
+	 *
+	 * @param integer $id Image ID.
+	 * @return array Image JSON.
+	 */
+	private function get_image_as_json( $id ) {
+		$path     = '/wp/v2/media/' . $id;
```

Comment:
Shouldn't this have the `edit` context set here?

Reply:
Thoughts @TimothyBJacobs

Reply:
Yeah that's a good point, that was brought over from my patch. We'll want to add it to the request object as a query param.